### ISSUE TITLE
Cell revisited during update

### ISSUE THREAD

**Issue by zyrolasting** [zyrolasting/kinda-ferpy#8]
I omit some printer output here, but I need to consider what to do about revisiting cells during an update.

Assume `X -> Y` means "`X` depends on `Y`". Here's an interaction where `c -> b -> a` and `c -> a`. If you update `a`, `c`'s body evaluates twice because an update propagates to all dependents even if one was already visited.

```
> (define a (% 1))
> (define b (% (displayln "b") (+ (a) 1)))
> (define c (% (displayln "c") (+ (a) (b))))
> (a 2)
c
b
c
2
```

This makes sense only to the extent that if any of `c` dependencies update, then `c` should update as well. Problem is, ~you can't update~ you probably shouldn't update `c` until ALL of `c`'s dependencies have finished updating. As it stands now, `c` is updated for every path through it on a single update.

**Comment by zyrolasting** [comment]
I think working with a topological sort will solve the problem, but I'm worried about how that impacts the interface. I don't want to compute a fresh topological sort every update since that's a lot of overhead when it's so easy to add nodes.

```
(transact
  (cell some-value #:traverse-by ordering))
```

**Comment by zyrolasting** [comment]
Alright, after some hacking here's what I'm going to try out: Instead of blindly propagating some new value of `u` to every dependent as before, I'm going to kick off a depth-first traversal starting from `u` and then only propagate change to the dependencies that the traversal visited exactly once. I don't like that every next update kicks off a new traversal, but I think this would produce the least surprising behavior. I'll also shim in a parameter to let users substitute their own procedure for deciding what cells update in response to `u`'s change if they need to optimize.

**Comment by zyrolasting** [comment]
Addressed in 97fad2bf7f2e7953d7bb26090bad543f09f70ab6

Took a bit of churn, but I landed on something that didn't add too much code.
